### PR TITLE
Fix three defects in named-character scoring, concrete anchors, and stream query parsing

### PR DESCRIPTION
**Pull request by Phazzie**
Two of the story-quality dimensions could only say yes, and the streaming route answered a malformed query with a response no client can read. All three were reproduced against the running modules before anything changed, and each fix has a test verified to fail on the old code.

## 1. Every sentence started a new "named character"

`scoreCharacterConsistency` collected `\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)?\b` over the whole story. That pattern matches the first word of every sentence as readily as a name, so this story — which has no cast in it at all —

> She opened a door and stepped through the cold hall.
> Rain fell hard. Blood pooled where the light could not reach.

reported `Named character count: 4`. The bonus caps at two names, so **every story of more than two sentences collected it in full**: the signal was a sentence count wearing a cast's name, and the dimension could not separate a story with characters from one without.

A capitalized word now counts only where it appears at least once away from a sentence boundary, a line start, or a `[Speaker]:` tag — where nothing but the word itself explains the capital.

*Stated limitation, and it is in the code:* a name a story only ever writes sentence-initially is now missed. That is the safe direction for an advisory signal, and dialogue speakers are already counted separately at three times the weight.

## 2. The concrete-anchor guard could never fire for the words it was written for

`extractConcreteAnchors` filtered tokens shorter than three characters out of the story **before** pairing them into phrases. The short words it dropped are exactly the ones its own `weakFirstTokens` list exists to reject — `a`, `an`, `my` — so those three entries were unreachable.

Worse, dropping a token welds its two non-adjacent neighbours into a phrase the prose never contained: "She opened **a** door" was reported as the specific anchor `opened door`, and "He carried **my** key" as `carried key`. A generic reference scored as a concrete one.

The pairing now runs over every token, so the guard reads the determiner that is actually there.

## 3. A repeated query parameter broke `/api/story/stream` two different ways

`?themes=romance&themes=dark` and `?creature=vampire&creature=witch` are ordinary URLs — a client appending to a `URLSearchParams` produces them without meaning to — and every runtime this route serves hands a repeat over as an array. Neither form reached the route's own validator:

- **`themes`** threw out of `themes.split(',')`, and the catch block answered the only way it knows: an SSE error frame. It writes that frame *before* `writeHead` has run, so the caller received a default 200 with no `text/event-stream` content type — a body no `EventSource` dispatches and no JSON parser reads — in place of the documented 400 `INVALID_INPUT`.
- **`creature`** did not throw at all. An array is truthy, so validation passed, the 200 and the `connected` frame went out, and the story service rejected the array mid-stream. The caller watched a healthy stream open and then fail, and the log recorded a 500 for what is a malformed request.

Query values now go through one reader that takes the first value, which is how `corsPolicy`, `security`, and the Story Lab blueprint parser already read every other possibly-repeated field. The catch block also branches on `res.headersSent`, so any failure *below* the stream is answered as an ordinary JSON error rather than as a frame nothing can read.

## Tests

Regression cases were added to the two existing suites, both already registered in `test:all`.

The heuristics get a **matched pair**: prose with no cast and no anchors must score as having none, and prose that names Mira and Lord Brine and anchors a witness shell and a blood oath must still score both — so the first assertion cannot be satisfied by a fix that simply scores nothing.

`FakeResponse` in the route-contracts test grew the `headersSent` that Node's own `ServerResponse` flips in `writeHead`. The route reads a real Node property; it was the double that did not model it.

Reverting each source file and re-running gives:

```
Error: sentence-initial words are not a cast (signals=["Named character count: 4"])
TypeError: themes.split is not a function
```

## Validation

- Vercel API `tsc --noEmit` sweep (the `slice-preflight.sh` invocation): clean.
- `npm run test:all`: passed, exit 0.
- **Not run:** the two Angular project type checks. `story-generator/node_modules` is absent in this environment, so both fail on `Cannot find module '@angular/core'` and `Cannot find type definition file for 'jasmine'` before reaching any source. Nothing here touches `story-generator/`.

`PR70_RECOVERY_CHANGELOG.md` and `LESSONS_LEARNED.md` are updated per the repo convention.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018VykbjXwz5uuX7FUYiq3H4

---
_Generated by [Claude Code](https://claude.ai/code/session_018VykbjXwz5uuX7FUYiq3H4)_

## Summary by Sourcery

Correct story-quality scoring and streaming query handling so malformed or generic input no longer produces misleading scores or unreadable responses.

Bug Fixes:
- Prevent sentence-initial words and generic references from being misclassified as named characters and concrete anchors.
- Handle repeated streaming query parameters consistently and return readable errors before or after an SSE stream is opened.

Enhancements:
- Add regression coverage for heuristic boundary cases, concrete-anchor filtering, repeated query parameters, and stream error handling.

Documentation:
- Record lessons learned and recovery details for the heuristic and streaming-route fixes.

Tests:
- Extend story quality and route contract suites with matched positive and negative regression cases.

### CODE REVIEW

File: api/_lib/story-lab/evaluation/storyQualityHeuristics.ts [hunk 1]
```diff
@@ -246,6 +245,74 @@ function extractDialogueSpeakers(dialogueLines: string[]): string[] {
     .filter((speaker): speaker is string => Boolean(speaker))));
 }
 
+const NAMED_CHARACTER_PATTERN = /\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)?\b/g;
```
[the remark this answers is not in the document]
Comment (Phazzie) [reply api/_lib/story-lab/evaluation/storyQualityHeuristics.ts]: Confirmed and fixed in 5c57f1c. `Then Mira pressed the blood oath.` produced no character signals at all — not the count, not the agency action.

The pattern now matches the whole run of consecutive capitalized words (`(?:\s+[A-Z][a-z]+)*`) and drops only the word actually sitting on the boundary, so the boundary is decided before the words are combined. `Then Mira` keeps `Mira`; `Then Lord Brine` keeps `Lord Brine`.

Covered by `testBoundaryRulesRejectOnlyTheBoundary` in `tests/story-quality-evals.test.ts`, which asserts both the count and the recovered agency action.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/_lib/story-lab/evaluation/storyQualityHeuristics.ts [hunk 2]
```diff
@@ -350,7 +417,14 @@ function extractConcreteAnchors(storyContent: string): string[] {
...
     .replace(/\s+/g, ' ');
-  const tokens = normalized.split(' ').filter(token => token.length > 2);
+  // Every token, not just the ones longer than two characters. The short words
+  // this used to drop are exactly the ones `weakFirstTokens` exists to catch —
+  // `a`, `an`, `my` — so the guard could never fire for them, and dropping a
+  // token also welded its two non-adjacent neighbours into a phrase that was
+  // never in the prose: "She opened a door" became the anchor `opened door`,
+  // and a generic reference was scored as a concrete one. Keeping the tokens
+  // lets the guard read the determiner that is actually there.
+  const tokens = normalized.split(' ').filter(Boolean);
```
[the remark this answers is not in the document]
Comment (Phazzie) [reply api/_lib/story-lab/evaluation/storyQualityHeuristics.ts]: Confirmed and fixed in 5c57f1c. `In court, she waited. By car, they crossed the city.` did emit `Specific anchors: in court, by car`.

Rather than extend the deny-list word by word, a first token of one or two characters is now rejected by length: no English word that short is a modifier that makes a noun specific, so this covers `in`, `by`, `to`, `of`, and `at` along with anything else of that size. `a`, `an`, and `my` came *out* of `weakFirstTokens` as a consequence — otherwise they'd be dead entries again, which was the original complaint about that list.

Covered in `testBoundaryRulesRejectOnlyTheBoundary`; the existing assertion that `witness shell`, `reef arch`, and `blood oath` still score keeps the rule from over-rejecting.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/_lib/story-lab/evaluation/storyQualityHeuristics.ts [hunk 3]
```diff
@@ -246,6 +245,74 @@ function extractDialogueSpeakers(dialogueLines: string[]): string[] {
...
 }
 
+const NAMED_CHARACTER_PATTERN = /\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)?\b/g;
+/**
+ * Punctuation that ends the sentence before a capitalized word, so the capital
+ * is explained by the position rather than by the word being a name. The colon
+ * is here for the `[Speaker]:` tags the generator writes, which put a capital
+ * after every one of them.
+ */
+const SENTENCE_END_PUNCTUATION = new Set(['.', '!', '?', ':', ';', '…', '—', '–']);
```
[the remark this answers is not in the document]
Comment (Phazzie) [reply api/_lib/story-lab/evaluation/storyQualityHeuristics.ts]: Confirmed and fixed in 392ca3e — your reproduction was exact (52 with no signals for the semicolon, 63 for the comma). The em dash had the same problem for the same reason.

The set now holds only the marks a capital can follow for positional reasons: `.`, `!`, `?`, the `:` of the `[Speaker]:` tags this generator writes, and `…`. The ellipsis stays because it genuinely can end a sentence, and for this dimension a missed name is the cheaper error than a common noun counted as a cast member.

The regression covers the semicolon, the em dash, and the comma together, since all three are mid-sentence marks that must leave the name standing.

---
_Generated by [Claude Code](https://claude.ai/code)_
